app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -0,0 +1,284 @@
+                                                        .build())
+                                        )
+                                        .setLimit(1)
+                                        .setOrder(Metric.Order.DESC)
+                                        .build()
+                        )).build());
+    }
+
+    private CompletionStage<MetricsQueryResponse> runRollupQuery(final LastDataPointMessage message) {
+        final MetricsQuery.Builder queryBuilder = new MetricsQuery.Builder();
+        final Metric.Builder metricBuilder = new Metric.Builder();
+        final String rollupMetricName = message.getMetricName() + message.getPeriod().getSuffix();
+        final RollupPeriod period = message.getPeriod();
+
+        final Instant lastDataPoint = message.getLastDataPointTime().orElse(Instant.EPOCH);
+        final Instant oldestBackfillPoint = period.recentEndTime(_clock.instant()).minus(period.periodCountToDuration(_maxBackFillPeriods));
+        if (lastDataPoint.isBefore(oldestBackfillPoint)) {
+            queryBuilder.setStartTime(oldestBackfillPoint);
+        } else {
+            queryBuilder.setStartTime(lastDataPoint);
+        }
+
+        queryBuilder.setEndTime(period.recentEndTime(_clock.instant()));
+
+        metricBuilder.setName(message.getMetricName());
+        if (!message.getTags().isEmpty()) {
+            metricBuilder.setGroupBy(ImmutableList.of(
+                    new MetricsQuery.GroupBy.Builder()
+                    .setName("tag").addOtherArg("tags", message.getTags())
+                    .build()
+            ));
+        }
+        metricBuilder.setAggregators(ImmutableList.of(
+                new Aggregator.Builder()
+                        .setName("merge")
+                        .setSampling(Optional.of(
+                                new Sampling.Builder()
+                                        .setValue(1)
+                                        .setUnit(message.getPeriod().getSamplingUnit())
+                                        .build()))
+                        .setAlignSampling(Optional.of(true))
+                        .setAlignEndTime(Optional.of(true))
+                        .build(),
+                new Aggregator.Builder()
+                        .setName("saveAs")
+                        .setOtherArgs(ImmutableMap.of("save_as", rollupMetricName))
+                        .build()
+        ));
+
+        return _kairosDbClient.queryMetrics(
+                queryBuilder.setMetrics(ImmutableList.of(metricBuilder.build()))
+                .build());
+    }
+
+    private final ActorRef _metricsDiscovery;
+    private final KairosDbClient _kairosDbClient;
+    private final int _maxBackFillPeriods;
+    private final FiniteDuration _fetchBackoff;
+    private final Clock _clock;
+    private List<RollupPeriod> _periodsInFlight = Collections.emptyList();
+    static final Object FETCH_METRIC = new Object();
```

Comment:
nit: convention is to have a blank line between non-static and static members.

Reply:
fixed